torch/onnx/_internal/fx/function_dispatcher.py

```diff
@@ -3,281 +3,199 @@
-):
-    # TODO(justinchuby): Improve how the table is constructed.
-    table: Dict[Union[torch._ops.OpOverload, Callable], str] = {}
-
-    # Some ops in `torch.ops.aten` are not discoverable through `dir(torch.ops.aten)`,
-    # but retrievable via explicit lookup.
-    # https://github.com/pytorch/pytorch/issues/99681
-    # This is a workaround to make sure we register ONNX symbolic functions for these.
-    onnx_supported_aten_lookup_table = [
-        k.split("::")[1] for k in _ATENLIB_FUNCTIONS.keys() if k.startswith("aten::")
-    ]
-
-    for op_namespace in (torch.ops.aten, torch.ops.prims):
-        attr_names = dir(op_namespace)
-        if op_namespace is torch.ops.aten:
-            attr_names += onnx_supported_aten_lookup_table
-        for attr_name in attr_names:
-            op_overload_packet = getattr(op_namespace, attr_name)
-            if not isinstance(op_overload_packet, torch._ops.OpOverloadPacket):
-                continue
-
-            exporter_look_up_key = op_overload_packet._qualified_op_name
-            if _ATENLIB_FUNCTIONS.get(exporter_look_up_key) is None:
-                # This aten op doesn't have ONNX exporter.
-                continue
-
-            for overload_name in op_overload_packet.overloads():
-                op_overload = getattr(op_overload_packet, overload_name)
-                # This line maps torch.ops.aten.add.Tensor, torch.ops.aten.add.Scalar, torch.ops.aten.add.out, etc
-                # to "aten::add". This means the exporter for "aten::add" is used for all overloads of "aten::add".
-                # This is applied to all ops under torch.ops.aten.
-                #
-                # TODO(wechi): in the future, we might want to write individual exporter for each overload, if,
-                # for example, they have different type promotion rules. If so, just map different overloads to
-                # different exporter keys.
-
-                table[op_overload] = op_overload_packet._qualified_op_name
-    return table
-
-
-# Dictionary that maps torch.ops.aten.* to exporter look up key; e.g.,
-# _OP_OVERLOAD_TO_EXPORTER_KEY_TABLE[torch.add.Tensor] is "aten::add".
-_OP_OVERLOAD_TO_EXPORTER_KEY_TABLE = _create_op_overload_to_exporter_key_table()
-_SYMINT_SYMFLOAT_BUILTIN_TO_EXPORTER_KEY_TABLE = {
-    operator.mul: "aten::mul",
-    operator.add: "aten::add",
-    operator.pow: "aten::pow",
-    operator.sub: "aten::sub",
+@runtime_checkable
+class WithDtype(Protocol):
+    @property
+    def dtype(self) -> torch.dtype:
+        ...
+
+
+class Dispatcher:
+    """
+    FX exporter dispatcher finds the best matched function for a given aten op.
+
+    # 1. Use torch.ops name to find the function
+        1a. Check aten:stft.center
+        2a. Check aten:stft (Fall back to default)
+
+    # 2. Find the best match among all overloaded functions.
+        2a. Type matched -> selected
+        2b. Type mismatched -> Nearest upcast/downcast
```

Comment:
Who is responsible for creating the cast nodes?

What is the design contract of this casting logic? What is the relationship of this with torch's "implicit type promotion", and "[implicit integral promotion](https://github.com/pytorch/pytorch/pull/100265#discussion_r1183025474)"?

Reply:
IMO, ONNX-Script torch_lib should provide type-matched function(overload) for each aten/prims op, so we should not have such case (type-mismatched).

Reply:
> Who is responsible for creating the cast nodes?
> 
> What is the design contract of this casting logic? What is the relationship of this with torch's "implicit type promotion", and "[implicit integral promotion](https://github.com/pytorch/pytorch/pull/100265#discussion_r1183025474)"?

The dispatcher will select the function. A separate type promotion logic will create the node.

> IMO, ONNX-Script torch_lib should provide type-matched function(overload) for each aten/prims op, so we should not have such case (type-mismatched).

We will when the onnx op does not support the input at all (asin etc.)

Reply:
This one can be put on hold. I think it's a "nice to have" feature.